nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -0,0 +1,1450 @@
+                }
+            };
+
+            json!({
+                "cleanup_report": cleanup_report,
+                "cleanup_err": cleanup_err,
+                "collection_report": collection_report,
+                "collection_err": collection_err,
+            })
+        }
+        .boxed()
+    }
+}
+
+// Takes a directory "dir", and zips the contents into a single zipfile.
+fn bundle_to_zipfile(dir: &Utf8TempDir) -> anyhow::Result<std::fs::File> {
+    let tempfile = tempfile()?;
+    let mut zip = ZipWriter::new(tempfile);
+
+    recursively_add_directory_to_zipfile(&mut zip, dir.path(), dir.path())?;
+
+    Ok(zip.finish()?)
+}
+
+fn recursively_add_directory_to_zipfile(
+    zip: &mut ZipWriter<std::fs::File>,
+    root_path: &Utf8Path,
+    dir_path: &Utf8Path,
+) -> anyhow::Result<()> {
+    // Readdir might return entries in a non-deterministic order.
+    // Let's sort it for the zipfile, to be nice.
+    let mut entries = dir_path
+        .read_dir_utf8()?
+        .filter_map(Result::ok)
+        .collect::<Vec<Utf8DirEntry>>();
+    entries.sort_by(|a, b| a.file_name().cmp(&b.file_name()));
+
+    for entry in &entries {
+        // Remove the "/tmp/..." prefix from the path when we're storing it in the
+        // zipfile.
+        let dst = entry.path().strip_prefix(root_path)?;
+
+        let file_type = entry.file_type()?;
+        if file_type.is_file() {
+            let opts = FullFileOptions::default();
+            let src = entry.path();
+
+            zip.start_file_from_path(dst, opts)?;
+            let mut reader = BufReader::new(std::fs::File::open(&src)?);
+
+            loop {
+                let buf = reader.fill_buf()?;
+                let len = buf.len();
+                if len == 0 {
+                    break;
+                }
+                zip.write_all(&buf)?;
+                reader.consume(len);
+            }
+        }
+        if file_type.is_dir() {
+            let opts = FullFileOptions::default();
+            zip.add_directory_from_path(dst, opts)?;
+            recursively_add_directory_to_zipfile(zip, root_path, entry.path())?;
+        }
+    }
+    Ok(())
+}
+
+async fn sha2_hash(file: &mut tokio::fs::File) -> anyhow::Result<ArtifactHash> {
+    let mut buf = vec![0u8; 65536];
+    let mut ctx = Sha256::new();
+    loop {
+        let n = file.read(&mut buf).await?;
+        if n == 0 {
+            break;
+        }
+        ctx.write_all(&buf[0..n])?;
+    }
+
+    let digest = ctx.finalize();
+    Ok(ArtifactHash(digest.as_slice().try_into()?))
+}
+
+// Identifies what we could or could not store within this support bundle.
+//
+// This struct will get emitted as part of the background task infrastructure.
+#[derive(Debug, Serialize, PartialEq, Eq)]
+struct CollectionReport {
+    bundle: SupportBundleUuid,
+
+    // True iff we could list in-service sleds
+    listed_in_service_sleds: bool,
+
+    // True iff the bundle was successfully made 'active' in the database.
```

Comment:
```suggestion
    // True if the bundle was successfully made 'active' in the database.
```

Reply:
I was using https://en.wikipedia.org/wiki/If_and_only_if , but I can make a less strong statement here if you prefer

Reply:
Ah, that makes more sense.  I should have realized that - sorry